src/paho/mqtt/publish.py

```diff
@@ -23,49 +23,45 @@
 import paho.mqtt as mqtt
 
 
-def _do_publish(c):
+def _do_publish(client):
     """Internal function"""
-    m = c._userdata.pop()
-    if type(m) is dict:
-        topic = m['topic']
-        try:
-            payload = m['payload']
-        except KeyError:
-            payload = None
-        try:
-            qos = m['qos']
-        except KeyError:
-            qos = 0
-        try:
-            retain = m['retain']
-        except KeyError:
-            retain = False
-    elif type(m) is tuple:
-        (topic, payload, qos, retain) = m
+
+    message = client._userdata.pop()
+
+    if isinstance(message, dict):
+        client.publish(message['topic'],
```

Comment:
@moisesguimaraes. By expanding the dictionary, all of the key/value pairs are passed to the publish method, not just those that are expected. Hence, it will throw an exception if there are unexpected inputs, which I think is the clearest behaviour.